### Import necessary libraries

In [ ]:
import numpy as np
import pandas as pd

import matplotlib.pyplot as plt 
import seaborn as sns
import pandas_profiling

from sklearn.model_selection import train_test_split
from sklearn.tree import DecisionTreeClassifier
from sklearn.metrics import accuracy_score, f1_score, recall_score, precision_score, confusion_matrix

from yellowbrick.classifier import ClassificationReport, ROCAUC

plt.style.use('ggplot')
pd.options.display.float_format = '{:,.2f}'.format
from IPython.core.display import display, HTML
display(HTML("<style>.container { width:95% !important; }</style>"))

/usr/local/lib/python3.6/dist-packages/sklearn/utils/deprecation.py:144: FutureWarning: The sklearn.metrics.classification module is  deprecated in version 0.22 and will be removed in version 0.24. The corresponding classes / functions should instead be imported from sklearn.metrics. Anything that cannot be imported from sklearn.metrics is now part of the private API.
  warnings.warn(message, FutureWarning)


In [ ]:
import warnings
warnings.filterwarnings('ignore')

In [ ]:
import string

In [ ]:
#AUC ROC curve
from sklearn.metrics import roc_auc_score
from sklearn.metrics import roc_curve

In [ ]:
#!pip install xlrd
#!pip install openpyxl

In [ ]:
import os
os.listdir('.')

['.config', 'output.xlsx', 'input.xlsx', 'ros_bag_data.bag', 'sample_data']

In [ ]:
import xlrd
df_in=pd.read_excel('input.xlsx',engine='openpyxl', header=None)
df_out=pd.read_excel('output.xlsx', engine='openpyxl',header=None)

In [ ]:
df_in.head()

0    1    2     3    4      5   ...     17   18    19    20   21    22
0 4.39 4.60 0.14 18.60 3.78 342.29  ... 636.41 4.89 19.53 18.11 1.47 73.67
1 4.02 5.16 0.50 16.31 2.51 186.37  ...  85.34 3.98 21.11 18.54 2.26 83.41
2 4.19 5.37 0.15 18.65 2.76 281.59  ... 547.28 2.72 18.39 21.32 1.57 60.68
3 4.11 4.68 0.69 19.84 2.49 853.99  ...  24.25 5.37 19.86 20.36 0.67 85.01
4 4.12 4.72 0.34 17.29 2.50 915.86  ... 787.17 5.57 21.74 21.18 2.96 35.75

[5 rows x 23 columns]

In [ ]:
df_out.head()

0     1    2     3
0  0.00  0.00 0.00  0.00
1 18.64 61.73 1.04 28.95
2 25.44 82.24 1.07 28.97
3  4.95 32.32 1.02 15.06
4  4.64 24.78 1.02 18.28

### Adding random colum names

In [ ]:
df_in.shape

(21920, 23)

In [ ]:
list(string.ascii_uppercase)[:df_in.shape[1]]

['A',
 'B',
 'C',
 'D',
 'E',
 'F',
 'G',
 'H',
 'I',
 'J',
 'K',
 'L',
 'M',
 'N',
 'O',
 'P',
 'Q',
 'R',
 'S',
 'T',
 'U',
 'V',
 'W']

In [ ]:
df_in.columns=list(string.ascii_uppercase)[:df_in.shape[1]]

In [ ]:
df_in.head()

A    B    C     D    E      F  ...      R    S     T     U    V     W
0 4.39 4.60 0.14 18.60 3.78 342.29  ... 636.41 4.89 19.53 18.11 1.47 73.67
1 4.02 5.16 0.50 16.31 2.51 186.37  ...  85.34 3.98 21.11 18.54 2.26 83.41
2 4.19 5.37 0.15 18.65 2.76 281.59  ... 547.28 2.72 18.39 21.32 1.57 60.68
3 4.11 4.68 0.69 19.84 2.49 853.99  ...  24.25 5.37 19.86 20.36 0.67 85.01
4 4.12 4.72 0.34 17.29 2.50 915.86  ... 787.17 5.57 21.74 21.18 2.96 35.75

[5 rows x 23 columns]

In [ ]:
df_out.shape

(21920, 4)

In [ ]:
df_out.columns=['out1','out2','out3','out4']

In [ ]:
df_out.head()

out1  out2  out3  out4
0  0.00  0.00  0.00  0.00
1 18.64 61.73  1.04 28.95
2 25.44 82.24  1.07 28.97
3  4.95 32.32  1.02 15.06
4  4.64 24.78  1.02 18.28

### Appending input and output

In [ ]:
df=df_in.append(df_out)

In [ ]:
df.head()

A    B    C     D    E      F  ...    V     W  out1  out2  out3  out4
0 4.39 4.60 0.14 18.60 3.78 342.29  ... 1.47 73.67   nan   nan   nan   nan
1 4.02 5.16 0.50 16.31 2.51 186.37  ... 2.26 83.41   nan   nan   nan   nan
2 4.19 5.37 0.15 18.65 2.76 281.59  ... 1.57 60.68   nan   nan   nan   nan
3 4.11 4.68 0.69 19.84 2.49 853.99  ... 0.67 85.01   nan   nan   nan   nan
4 4.12 4.72 0.34 17.29 2.50 915.86  ... 2.96 35.75   nan   nan   nan   nan

[5 rows x 27 columns]

In [ ]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 43840 entries, 0 to 21919
Data columns (total 27 columns):
 #   Column  Non-Null Count  Dtype  
---  ------  --------------  -----  
 0   A       21920 non-null  float64
 1   B       21920 non-null  float64
 2   C       21920 non-null  float64
 3   D       21920 non-null  float64
 4   E       21920 non-null  float64
 5   F       21920 non-null  float64
 6   G       21920 non-null  float64
 7   H       21920 non-null  float64
 8   I       21920 non-null  float64
 9   J       21920 non-null  float64
 10  K       21920 non-null  float64
 11  L       21920 non-null  float64
 12  M       21920 non-null  float64
 13  N       21920 non-null  float64
 14  O       21920 non-null  float64
 15  P       21920 non-null  float64
 16  Q       21920 non-null  float64
 17  R       21920 non-null  float64
 18  S       21920 non-null  float64
 19  T       21920 non-null  float64
 20  U       21920 non-null  float64
 21  V       21920 non-null  float64
 22

- All are float values
- there is no null values

In [ ]:
df.describe().T

count   mean    std   min    25%    50%    75%    max
A    21,920.00   4.30   0.23  3.90   4.10   4.30   4.50   4.70
B    21,920.00   4.90   0.29  4.40   4.66   4.90   5.15   5.40
C    21,920.00   0.50   0.29  0.00   0.25   0.50   0.75   1.00
D    21,920.00  16.98   1.73 14.00  15.48  16.97  18.46  20.00
E    21,920.00   3.24   1.01  1.50   2.36   3.23   4.11   5.00
F    21,920.00 497.89 287.62  0.02 250.10 499.00 744.63 999.94
G    21,920.00 502.90 288.26  0.13 254.59 504.84 752.80 999.99
H    21,920.00   3.98   1.73  1.00   2.47   3.97   5.49   7.00
I    21,920.00  19.99   1.16 18.00  18.98  20.00  20.98  22.00
J    21,920.00  20.00   1.16 18.00  19.01  20.00  21.00  22.00
K    21,920.00   4.10   0.18  3.90   3.95   4.05   4.21   4.70
L    21,920.00  51.04  28.26  2.00  26.64  51.01  75.36 100.00
M    21,920.00   5.09   2.87  0.10   2.63   5.11   7.59  10.00
N    21,920.00   0.50   0.29  0.00   0.25   0.50   0.75   1.00
O    21,920.00  16.99   1.73 14.00  15.49  17.00  18.47  20.00
P    21,920.00   3.25   1.01  1.50   2.39   3.24   4.13   5.00
Q    21,920.00 499.18 291.03  0.03 246.18 498.07 752.44 999.94
R    21,920.00 500.81 287.76  0.06 252.88 500.48 748.80 999.99
S    21,920.00   4.01   1.73  1.00   2.52   4.01   5.50   7.00
T    21,920.00  20.01   1.15 18.00  19.01  20.01  21.01  22.00
U    21,920.00  20.00   1.15 18.00  19.01  20.00  20.99  22.00
V    21,920.00   1.90   1.03 -0.53   1.02   1.90   2.77   3.89
W    21,920.00  50.92  28.48  2.00  26.32  50.58  75.64 100.00
out1 21,920.00   9.84   6.44  0.00   4.77   8.21  14.25  26.69
out2 21,920.00  40.90  19.03  0.00  28.03  37.27  52.02  89.06
out3 21,920.00   1.03   0.21  0.00   1.02   1.04   1.05   8.95
out4 21,920.00  21.39   7.19  0.00  15.89  23.00  28.16  29.49

- data is not skewed. Mean and median are very similiar for each column

In [ ]:
for i in df.columns:
    print(i)
    sns.distplot(df[i])
    plt.show()

A
B
C
D
E
F
G
H
I
J
K
L
M
N
O
P
Q
R
S
T
U
V
W
out1
out2
out3
out4


In [ ]:
plt.figure(figsize=[20,30])
sns.heatmap(df.corr(),annot=True)

- A, K are correlated
- P and V are highly negatively correlated
- Columns should be independent to each other before applying ML algorithms. For taht we need to either drop columns or we have to do dimensinality reduction
- As we don't the real column names, I prefer dimensionality reduction techniques

In [ ]:
from sklearn.preprocessing import StandardScaler 
from sklearn.decomposition import PCA 
from sklearn.linear_model import LinearRegression 
from sklearn.tree import DecisionTreeRegressor
from sklearn.pipeline import Pipeline
from sklearn.model_selection import GridSearchCV
from sklearn.ensemble import RandomForestRegressor

In [ ]:
from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(df_in, df_out, test_size = .30, random_state=0)

In [ ]:
pipe_lr = Pipeline([('scl', StandardScaler()), ('pca', PCA(n_components=23)), ('clf', LinearRegression())]) 
pipe_lr.fit(X_train, y_train) 
print('Test Accuracy: %.3f' % pipe_lr.score(X_test, y_test))
y_pred = pipe_lr.predict(X_test)
from sklearn import metrics
print('Mean Absolute Error:', metrics.mean_absolute_error(y_test, y_pred))
print('Mean Squared Error:', metrics.mean_squared_error(y_test, y_pred))
print('Root Mean Squared Error:', np.sqrt(metrics.mean_squared_error(y_test, y_pred)))

Test Accuracy: 0.688
Mean Absolute Error: 3.3244151863310814
Mean Squared Error: 35.761560759973
Root Mean Squared Error: 5.980097052721887


In [47]:
for i in range(4):
    print('Root Mean Squared Error:', np.sqrt(metrics.mean_squared_error(y_test.iloc[:,i], y_pred[:,i])))

Root Mean Squared Error: 3.20948663723674
Root Mean Squared Error: 10.38131292786422
Root Mean Squared Error: 0.2037104505473943
Root Mean Squared Error: 4.99322365925941


### neural networks

In [ ]:
from sklearn.neural_network import MLPRegressor   


In [ ]:
regr = MLPRegressor(random_state=1, max_iter=500).fit(X_train, y_train)
regr.score(X_test, y_test)

0.651467976891008

In [ ]:
y_pred_regr = regr.predict(X_test)

for i in range(4):
    print('Root Mean Squared Error:', np.sqrt(metrics.mean_squared_error(y_test.iloc[:,i], y_pred_regr[:,i])))

Root Mean Squared Error: 3.469293726829775
Root Mean Squared Error: 11.009550810021487
Root Mean Squared Error: 0.5283404906178082
Root Mean Squared Error: 5.136858144537245


In [44]:
pipe_MLP = Pipeline([('scl', StandardScaler()), ('pca', PCA(n_components=22)), ('clf', MLPRegressor())]) 
param_grid_MLP = {'clf__hidden_layer_sizes':list(range(50,100,20)),'clf__solver':['sgd','adam'],'clf__learning_rate':['constant','adptive']} 
grid_MLP = GridSearchCV( pipe_MLP , param_grid = param_grid_MLP, cv = 5) 
grid_MLP.fit( X_train, y_train) 
print(" Best parameters: ", grid_MLP.best_params_) 

 Best parameters:  {'clf__hidden_layer_sizes': 90, 'clf__learning_rate': 'constant', 'clf__solver': 'adam'}


In [46]:
print(grid_MLP.score(X_test, y_test))
y_pred_MLP = grid_MLP.predict(X_test)
for i in range(4):
    print('Root Mean Squared Error:', np.sqrt(metrics.mean_squared_error(y_test.iloc[:,i], y_pred_MLP[:,i])))

0.7774441765193517
Root Mean Squared Error: 2.6144413252729195
Root Mean Squared Error: 9.224769136697573
Root Mean Squared Error: 0.2061206206326614
Root Mean Squared Error: 3.184085046488475


### Conclusion
#### Linear Regression
* Coeficient of determination R^2 of prediction: 0.688
* Root Mean Squared Error of first output: 3.20948663723674
* Root Mean Squared Error second output: 10.38131292786422
* Root Mean Squared Error third output: 0.2037104505473943
* Root Mean Squared Error forth output: 4.99322365925941


#### Neural networks


*   Coeficient of determination R^2 of prediction: 0.7774441765193517
*   Root Mean Squared Error of first output: 2.614441325272919
*   Root Mean Squared Error of second output: 9.224769136697573
*   Root Mean Squared Error of third output: 0.2061206206326614
*   Root Mean Squared Error of forth output: 3.184085046488475


